File: fly-promo-publisher-flyWU/src/main/java/io/github/riojano0/flypromoflyWU/service/FlyPromoService.java
package io.github.riojano0.flypromoflyWU.service;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Random;

import io.github.riojano0.flypromoflyWU.model.CountryProvider;
import io.github.riojano0.flypromoflyWU.model.FlyWuFlyPromo;
import io.github.riojano0.flypromokafka.core.model.Country;
import io.github.riojano0.flypromokafka.core.model.FlyPromo;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.kafka.support.SendResult;
import org.springframework.stereotype.Service;
import org.springframework.util.concurrent.ListenableFuture;

import lombok.extern.slf4j.Slf4j;

@Service
@Slf4j
public class FlyPromoService {

   private static final Map<Integer, BigDecimal[]> PROMOTIONAL_PRICES =
         Map.of(
               0, new BigDecimal[] { BigDecimal.valueOf(100), BigDecimal.valueOf(10) },
               1, new BigDecimal[] { BigDecimal.valueOf(100), BigDecimal.valueOf(30) },
               2, new BigDecimal[] { BigDecimal.valueOf(100), BigDecimal.valueOf(50) },
               3, new BigDecimal[] { BigDecimal.valueOf(100), BigDecimal.valueOf(90) }
               );
   private static final Random RANDOM = new Random();

   private final KafkaTemplate<String, FlyPromo> kafkaTemplate;

   @Autowired
   public FlyPromoService(@Qualifier("kafkaFlyPromoTemplate") KafkaTemplate<String, FlyPromo> kafkaTemplate) {
      this.kafkaTemplate = kafkaTemplate;
   }

   public FlyWuFlyPromo generateFlyPromo() {
      FlyWuFlyPromo flyWuFlyPromo = new FlyWuFlyPromo();
      CountryProvider[] values = CountryProvider.values();
      Country toCountry = values[RANDOM.nextInt(values.length)];
      flyWuFlyPromo.setToCountry(toCountry);

      List<Country> fromCountries = new ArrayList<>();
      do {
         for (int i = 0; i < 2; i++) {
            Country fromCountry = values[RANDOM.nextInt(values.length)];
            if (fromCountry != toCountry) {
               fromCountries.add(fromCountry);
            }
         }
      } while (fromCountries.isEmpty());
      flyWuFlyPromo.setFromCountries(fromCountries);

      BigDecimal[] bigDecimals = PROMOTIONAL_PRICES.get(RANDOM.nextInt(PROMOTIONAL_PRICES.size()));

      flyWuFlyPromo.setOriginalPrice(bigDecimals[0]);
      flyWuFlyPromo.setPromotionalPrice(bigDecimals[1]);
      flyWuFlyPromo.setStock(10);
      LocalDateTime now = LocalDateTime.now();
      flyWuFlyPromo.setStartDate(now);
      flyWuFlyPromo.setEndDate(now.plusDays(1));

      ListenableFuture<SendResult<String, FlyPromo>> future = kafkaTemplate.send("fly-promo", flyWuFlyPromo);

      future.addCallback(
            result -> log.info("Complete Fly Promo"),
            result -> log.error("Unable to send promo", result.getCause()));

      return flyWuFlyPromo;
   }

}

File: fly-promo-publisher-fly2EU/src/main/kotlin/io/github/riojano0/flypromofly2EU/service/FlyPromoService.kt
package io.github.riojano0.flypromofly2EU.service

import io.github.riojano0.flypromokafka.core.model.Country
import io.github.riojano0.flypromokafka.core.model.FlyPromo
import io.github.riojano0.flypromofly2EU.model.CountryProvider
import io.github.riojano0.flypromofly2EU.model.Fly2EUFlyPromo
import lombok.extern.slf4j.Slf4j
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.time.LocalDateTime
import java.util.Random


@Service
@Slf4j
class FlyPromoService
@Autowired constructor(
        @Qualifier("kafkaFlyPromoTemplate") val kafkaTemplate: KafkaTemplate<String, FlyPromo>) {

    private final val log: Logger = LoggerFactory.getLogger(FlyPromoService::class.java)
    private final val random: Random = Random()
    private final val promotionalPrices = mutableMapOf(
            0 to listOf<BigDecimal>(BigDecimal.valueOf(1000), BigDecimal.valueOf(750)),
            1 to listOf<BigDecimal>(BigDecimal.valueOf(1000), BigDecimal.valueOf(500)),
            2 to listOf<BigDecimal>(BigDecimal.valueOf(1000), BigDecimal.valueOf(450)),
            3 to listOf<BigDecimal>(BigDecimal.valueOf(1000), BigDecimal.valueOf(950))
    )

    fun generateFlyPromo() : FlyPromo {
        val fly2EuFlyPromo = Fly2EUFlyPromo()
        val values = CountryProvider.values()
        val toCountry = values[random.nextInt(values.size)]
        fly2EuFlyPromo.toCountry = toCountry

        val countries = arrayListOf<Country>()

        do {
            for (i in 0..2) {
                val country = values[random.nextInt(values.size)]
                if (country != toCountry) {
                    countries.add(country)
                }
            }
        } while (countries.isEmpty())
        fly2EuFlyPromo.fromCountries = countries;

        val bigDecimals = promotionalPrices[random.nextInt(promotionalPrices.size)]

        fly2EuFlyPromo.originalPrice = bigDecimals?.get(0)
        fly2EuFlyPromo.promotionalPrice = bigDecimals?.get(1)
        fly2EuFlyPromo.stock = 10
        val now = LocalDateTime.now()
        fly2EuFlyPromo.startDate = now
        fly2EuFlyPromo.endDate = now.plusDays(1)

        val listenableFuture = kafkaTemplate.send("fly-promo", fly2EuFlyPromo)
        listenableFuture.addCallback(
                { log.info("Complete Fly Promo") },
                { result -> log.error("Unable to send promo", result.cause) }
        )

        return fly2EuFlyPromo
    }

}